tracing: switch from OpenTracing to OpenTelemetry

andreimatei opened the pull request:
The industry is converging on OpenTelemetry as the interface for
defining traces and metrics, and also as the messaging formats for
collecting these traces and metrics into analysis tools. Up until this
patch, CRDB has used the older OpenTracing interfaces for years for
producing traces for 3rd-party tools (for metrics we only support
exporting them through a Prometheus scraping interface, and this patch
doesn't do anything new with metrics). In the meantime, OpenTracing has
merged with another project - OpenCensus - to form OpenTelemetry (otel).
All the observability world has been standardizing on otel (e.g. Jaeger
[1]; all the collectors we currently support work with OpenTelemetry
natively, with the exception of Zipkin that works through the
OpenTelemetry Collector [2]).

A nice thing about OpenTelemetry is that, besides defining the
interfaces that the instrumented code needs to use to define its traces,
it also provides a canonical tracer and defines standard formats (grpc
and http) for that tracer to export its data to any trace collector.
With OpenTracing, we had to a tracer for each collector that we wanted
to export to (we imported Lightstep, Zipkin and DataDog). Now, we no
longer have dependencies to these projects.

The OpenTelemetry project also comes with a canonical trace collector
([2]) that knows how to forward traces to any number of tools at once.
We did not have this flexibility in CRDB before; one could only export
to one trace collector at a time. The trace collector also knows how to
export traces to analysis tools that don't yet speak the native otlp
protocols; it supports many many tools.

This modernization also simplified our code, letting us delete confusing
code that dealt with injecting/extracting the OpenTracing tracing
context into RPC "carriers". This operations used to depend on the type
of OpenTracing tracer being used in an opaque way. Now, there's a single
OpenTelemetry tracer to speak of. Generally, you'll find that much is
simplified with this patch.

This patch introduces the trace.opentelemetry.collector cluster setting,
replacing all the collector-specific flags we had before. It takes the
tcp address of a collector speaking the grpc otlp protocol. There's no
backwards compatibility.

[1] https://www.jaegertracing.io/docs/1.21/opentelemetry/
[2] https://github.com/open-telemetry/opentelemetry-collector

Release note (general change): CRDB now supports exporting operation
traces to standard OpenTelemetry-compatible tools through the
trace.opentelemetry.collector cluster setting. Support for exporting to
DataDog, Lightstep and Zipkin through other interfaces has been retired;
all these tools can use OpenTelemetry data.

Closes #57908

andreimatei committed:
tracing: switch from OpenTracing to OpenTelemetry

The industry is converging on OpenTelemetry as the interface for
defining traces and metrics, and also as the wire formats for collecting
these traces and metrics into analysis tools (the OTLP protocol). Up
until this patch, CRDB has used the older OpenTracing interfaces for
years for producing traces for 3rd-party tools (for metrics we only
support exporting them through a Prometheus scraping interface, and this
patch doesn't do anything new with metrics). In the meantime,
OpenTracing has merged with another project - OpenCensus - to form
OpenTelemetry (otel).

A nice thing about OpenTelemetry is that, besides defining the
interfaces that the instrumented code needs to use to define its traces,
it also provides a canonical tracer and defines standard formats (grpc
and http) for that tracer to export its data to any trace collector.
With OpenTracing, we had to a tracer for each collector that we wanted
to export to (we imported Lightstep, Zipkin and DataDog). Now, we no
longer have dependencies to these projects.

The OpenTelemetry project also comes with a canonical trace collector
([2]) that knows how to forward traces to any number of tools at once.
We did not have this flexibility in CRDB before; one could only export
to one trace collector at a time. The trace collector also knows how to
export traces to analysis tools that don't yet speak the native otlp
protocols; it supports many many tools.

This patch removes support for OpenTracing tracers and adds the otel
tracer and a couple of exporters. For outputting traces to 3rd party
systems, we used to embed the Datadog, Lightstep and Zipkin opentracing
tracers; these are all gone. In their stead, we now use the (one and
only) otel tracer, plus one or more of the otlp, jaeger and zipkin
*exporters* (the otel exporters are small libraries that plug into the
tracer, with a smaller footprint than the old opentracing tracers).

We now have the following many options for exporting traces to 3rd
parties (the otel project provides flexibility to a fault):

1. We can export to anybody that speaks the OTLP protocolo. Notably, we
   can export to the OpenTelemetry Collector protocol. The Collector can
   be configured to forward traces to all imaginable tools (and also to
   many tools at once).
   Some (increasingly many) tools also speak OTLP natively. For example,
   Jaeger does, but for now only when properly configured. For example the
   docker image `jaegertracing/opentelemetry-all-in-one` has the proper
   configuration. This is not the default Jaeger image, and it hasn't
   recently been updated to new Jaeger versions, pending "Jaeger 2"
   developments.

2. We can export using native protocols to Jaeger and Zipkin, by using
   their own exporters. Before this patch, we could natively export only to
   Zipkin, but Jaeger could be configured to accept the Zipkin protocol.
   Now, we speak native Jaeger so that configuration is no longer required.

Option 1) is used through the new `trace.opentelemetry.collector`
cluster setting.
Option 2) is used as follows:
- For Zipkin, the existing cluster setting `trace.zipkin.collector`.
- For Jaeger, the new cluster setting `trace.jaeger.agent`. Jaeger has
  different ports/procols for communication between an app and an
  "agent" versus an agent and a "collector", hence the setting name. Note
  that this new setting can be used instead of the previous way of
  outputting to Jaeger, which was by using the `trace.zipkin.collector`
  setting and configuring Jaeger with the
  `COLLECTOR_ZIPKIN_HOST_PORT=9411` env var. This configuration still
  works, though.

The cluster settings trace.lightstep.token, trace.datadog.agent and
trace.datadog.agent are removed. Lightstep speaks native OTLP. For
DataDog, the Collector can be used.

This modernization also simplified our code, letting us delete confusing
code that dealt with injecting/extracting the OpenTracing tracing
context into RPC "carriers". This operations used to depend on the type
of OpenTracing tracer being used in an opaque way. Now, there's a single
OpenTelemetry tracer to speak of. Generally, you'll find that much is
simplified with this patch.

[1] https://www.jaegertracing.io/docs/1.21/opentelemetry/
[2] https://github.com/open-telemetry/opentelemetry-collector

Release note (general change): CRDB now supports exporting operation
traces to OpenTelemetry-compatible tools using the OTLP protocol through
the trace.opentelemetry.collector cluster setting.

CRDB now support exporting traces to a Jaeger agent through the
new trace.jaeger.agent cluster setting. Exporting to Jaeger was
previously possibly by configuring the Jaeger agent to accept Zipkin
traces and using the trace.zipkin.collector cluster setting; this
configuration is no longer required.

Support for exporting to DataDog and Lightstep through other interfaces
has been retired; these tools can use OpenTelemetry data.  The cluster
settings trace.lightstep.token, trace.datadog.agent, and
trace.datadog.project have been deprecated; they no longer have any
effect.

Closes #57908

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65599)
<!-- Reviewable:end -->

tbg reviewed (commented):
Haven't reviewed the change in any detail, but doing this SGTM. A few things to look out for:

1. Irfan just did perf work on this package, would be good to see if we're regressing in any meaningful way here.
2. How do the old cluster settings and the new ones interact? I scrolled through the diff but didn't see the old ones getting removed (I possibly missed that). Be specific in the release note so that folks know how to change things.

irfansharif reviewed (approved):
The changes here LGTM, though it'd be worth having Rima + obs peeps also take a look first to build familiarity. To check against regressions I'd just run BenchmarkTracing to see that there aren't any unexpected allocations around extracting/injecting meta. Looking at the code, I don't think there should be.

If I understand correctly, instead of feeding to datadog/zipkin/jaeger directly, we'd now push out trace data to an otel collector, that would in turn forward to all these other tools? Or some of these tools expose an endpoint that lets it talk using the otel APIs? It'd be worth spelling it out a bit more in the release note. We also have a few wiki pages that walk through how to use zipkin+friends, and it'd be worth updating those. I think they're frequently read.

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-MaZLRjjDFvXUV8hbF7j:bjncgw2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @rimadeodhar)

___
*[docs/generated/settings/settings-for-tenants.txt, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYqx6l56__mkrU1TFc:-MaYqx6l56__mkrU1TFd:bdogl5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/docs/generated/settings/settings-for-tenants.txt#L108)):*
> ```vimhelp
> timeseries.storage.resolution_30m.ttl	duration	2160h0m0s	the maximum age of time series data stored at the 30 minute resolution. Data older than this is subject to deletion.
> trace.debug.enable	boolean	false	if set, traces for recent requests can be seen at https://<ui>/debug/requests
> trace.opentelemetry.collector	string		address of an OpenTelemetry trace collector to receive traces using the otel gRPC protocol
> ```

Not this PR, but are tenants allowed to define their own collectors? Where do we ensure traces from other tenants don't get collected in this tenant's collector?
___
*[pkg/util/tracing/grpc_interceptor.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZHOKCBO7IshUjlTjO:-MaZHOKCBO7IshUjlTjP:b-tt6fo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/grpc_interceptor.go#L81)):*
> ```Go
> // setSpanTags sets one or more tags on the given span according to the
> // error.
> func setSpanTags(sp *Span, err error) {
> ```

Looks like all we're using this for is to set the error status? Good time to improve the naming for setSpanTags to be a bit more specific. 

[nit] I think I'd also prefer not exporting sp.SetStatus, is it something we expect callers of pkg/tracing to ever use? If not, maybe it's worth just inlining it here.
___
*[pkg/util/tracing/grpc_interceptor.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYs8vn6X-EhuqkAvdX:-MaYs8vn6X-EhuqkAvdY:b-t9f5v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/grpc_interceptor.go#L321)):*
> ```Go
> 	finishFunc := func(err error) {
> 		//  Since we have multiple code paths that could concurrently call
> 		//  `finishFunc`, we need to add some sort of synchronization to guard
> ```

[nit] Double spaces after //
___
*[pkg/util/tracing/span.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZIbxvB7fHxveOjsOe:-MaZIbxvB7fHxveOjsOf:b-f0mme2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span.go#L222)):*
> ```Go
> 
> // SetStatus sets an RPC response code on the span.
> func (sp *Span) SetStatus(code codes.Code, msg string) {
> ```

See comment above, I don't think it's that useful to export this outside this package, and maybe worth inlining at caller.
___
*[pkg/util/tracing/span_inner.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYsu-dAXfiZB65-bKo:-MaYsu-dAXfiZB65-bKp:b-1w0c7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_inner.go#L150)):*
> ```Go
> 	}
> 
> 	var shadowCtx oteltrace.SpanContext
> ```

s/shadowCtx/otelCtx?
___
*[pkg/util/tracing/span_options.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYtd2G-DvtUwtZ--A6:-MaYtd2G-DvtUwtZ--A7:bycirr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_options.go#L31)):*
> ```Go
> 	// ChildOfRef means that the parent span will wait for the child span's
> 	// termination.
> 	ChildOfRef SpanReferenceType = iota
> ```

Do these symbols need to be exported?
___
*[pkg/util/tracing/span_options.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYu0BU3884rqgRVnUJ:-MaYu0BU3884rqgRVnUK:bca4lbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_options.go#L86)):*
> ```Go
> 
> // otelContext returns information about the OpenTelemetry parent span. If there
> // is a parent, open of a Span (if the parent is local) or a SpanContext (if the
> ```

"open of a Span"?
___
*[pkg/util/tracing/tracer.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYuUiJDTyVrgwduQny:-MaYuUiJDTyVrgwduQnz:b-wvwcnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L86)):*
> ```Go
> 	"trace.opentelemetry.collector",
> 	"address of an OpenTelemetry trace collector to receive traces using the otel gRPC protocol",
> 	envutil.EnvOrDefaultString("COCKROACH_OTLP", ""),
> ```

What's LP?
___
*[pkg/util/tracing/tracer.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYuYQ-9lRGWeQFPXqb:-MaYuYQ-9lRGWeQFPXqc:biqod1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L89)):*
> ```Go
> ).WithPublic()
> 
> // Tracer provides suppoort for tracing requests. It supports:
> ```

s/suppoort/support (+ nit: rephrase to deduplicate support?)
___
*[pkg/util/tracing/tracer.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYw-Ti-AlncHm2RqfN:-MaYw-Ti-AlncHm2RqfO:b-s6z28n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L253)):*
> ```Go
> // setShadowTracer sets the OpenTelemetry tracer to use as a "shadow tracer". A
> // nil value means that no shadow tracer will be used.
> func (t *Tracer) setShadowTracer(tr oteltrace.Tracer) {
> ```

[nit, throughout] If there's only one kind of shadow tracer now, the OpenTelemetry kind, should we remove all code references to "shadow" tracers? (in favor of otel{Tracers,Span,Ctx,...})
___
*[pkg/util/tracing/tracer.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYxnpXD_fCg5m7oAas:-MaYxnpXD_fCg5m7oAat:bx8296e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L348)):*
> ```Go
> 		shadowTr := t.getShadowTracer()
> 
> 		// If both are set and don't agree, ignore shadow tracer
> ```

This comment can go away.
___
*[pkg/util/tracing/tracer.go, line 574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZFDjL0uvWpV0fzID9:-MaZFDjL0uvWpV0fzIDA:b1ymtat) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L574)):*
> ```Go
> 			traceID, err = strconv.ParseUint(v, 16, 64)
> 			if err != nil {
> 				return errors.Errorf("invalid trace id: %x", traceID)
> ```

Here and below, if there is an error, I think `traceID` would be the zero value. Also below, our error msg captures the traceID, not the spanID. I think printing out the err + value be better.
___
*[pkg/util/tracing/tracer.go, line 631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYyOqt-V8ys3hpk8E-:-MaYyOqt-V8ys3hpk8E0:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L631)):*
> ```Go
> 	var shadowCtx oteltrace.SpanContext
> 	if otelTraceID.IsValid() && otelSpanID.IsValid() {
> 		shadowCtx = shadowCtx.WithRemote(true).WithTraceID(otelTraceID).WithSpanID(otelSpanID)
> ```

Nice.


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (commented):
The changes look good (admittedly, don't have a lot of context to provide an in depth review but overall seems good). I had the same question as Irfan on the implications of this change on whether the trace data will be pushed to an otel collector or does it work in some other way?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-Ma_YRo5A1rnVswNXslB:b-bxplly)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @irfansharif)

___
*[pkg/util/tracing/tracer.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYuUiJDTyVrgwduQny:-MambTCE2aaw2PYHjt0e:b-obi20g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L86)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

What's LP?
</blockquote></details>

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md

I also thought this wasn't quite the right name for this env var. It should be something like COCKROACH_OTLP_COLLECTOR


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
This one should be updated: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73171339/Tracing+logs+with+Jaeger+and+Zipkin

Frankly I'm not sure how this all works out of the box with jaeger-all-in-one. I think we need to switch to this all-in-one image: https://hub.docker.com/r/jaegertracing/opentelemetry-all-in-one/

If you're using zipkin, it's probably even more complicated. Honestly I'm not sure we should pull support for the other endpoints unless it's clear how they would be connected.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-ManHbQ2AhB0aqvDoKLn:b-itj8tt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @irfansharif)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Wondering how https://github.com/cockroachdb/cockroach/pull/65844 would look like with this PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-Manc2oKCqMznE7geGdw:b1y2ljw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @irfansharif)



<!-- Sent from Reviewable.io -->

andreimatei committed:
util/tracing: tighten span tags

A span's tag were needlessly generic (interface{}). This doesn't work
great with OpenTelemetry, which wants tags ("Attributes", in otel
parlance) to have a statically known type. There is an `Attribute.Any`,
but it's deprecated (and our linter complains).
This patch switches to using otel's Attribute type, which has good type
safety. It also avoids some boxing allocations.

Release note: None

andreimatei committed:
tracing: remove unused span opt

SpanOptions.Tags was unused since the WithSpans option was removed in
eba03c4243693df39c0152680c13c8d875428b27.

Release note: None

andreimatei reviewed (commented):
I think this is in really good shape now. @irfansharif perhaps a look at the last commit.

I've polished this some more. I've made it work natively with all Jaeger versions, and also with Zipkin, without needing to go through the otel collector. This works because I've included the Jaeger and Zipkin *exporters* into the binary, in addition to the generic "otlp" exporter. More in the commit message.
So the old setting `trace.zipkin.collector` works again, and there's now also a `trace.jaeger.agent`. Jaeger no longer needs to be running with the Zipkin import endpoint configured (the `COLLECTOR_ZIPKIN_HOST_PORT=9411` env var that we've been passing to to it).

> This one should be updated: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73171339/Tracing+logs+with+Jaeger+and+Zipkin

Yes, I will update it when merging.

> Frankly I'm not sure how this all works out of the box with jaeger-all-in-one. I think we need to switch to this all-in-one image: https://hub.docker.com/r/jaegertracing/opentelemetry-all-in-one/

As I said above, I've made all jaeger images work now directly, not just the `opentelemetry-all-in-one` one.

> If I understand correctly, instead of feeding to datadog/zipkin/jaeger directly, we'd now push out trace data to an otel collector, that would in turn forward to all these other tools? Or some of these tools expose an endpoint that lets it talk using the otel APIs?

The integrations work in all ways imaginable:
1. There is a collector maintained under the OpenTelemetry project that knows how to fwd the traces to any tool. You export to that collector using the "otlp" protocol.
2. Some tools have native support for ingesting over otlp. For example, Jaeger does, but only when properly configured (like the `jaegertracing/opentelemetry-all-in-one` does).
3. For most tools that don't ingest directly over otlp, you can also embed native "exporters" in your code. These are much smaller library than the full tracers that you had to import with `opentracing` (i.e. your code still uses the one and only otel `Tracers`, but you can plug these exporters into it). I've now included importers for Jaeger and Zipkin, for maximum convenience.

I've ran `BenchmarkTracing`; no difference.

```
name                                            old time/op    new time/op    delta
Tracing/Cockroach/tracing=f/Scan1-32               195µs ± 7%     184µs ± 5%    ~     (p=0.056 n=5+5)
Tracing/Cockroach/tracing=f/Insert-32              293µs ± 3%     284µs ± 2%    ~     (p=0.222 n=5+5)
Tracing/Cockroach/tracing=t/Scan1-32               204µs ± 6%     216µs ± 6%    ~     (p=0.095 n=5+5)
Tracing/Cockroach/tracing=t/Insert-32              303µs ± 3%     304µs ± 2%    ~     (p=0.841 n=5+5)
Tracing/MultinodeCockroach/tracing=f/Scan1-32      394µs ± 5%     393µs ±11%    ~     (p=0.690 n=5+5)
Tracing/MultinodeCockroach/tracing=f/Insert-32     498µs ± 5%     487µs ± 7%    ~     (p=0.690 n=5+5)
Tracing/MultinodeCockroach/tracing=t/Scan1-32      436µs ± 4%     443µs ± 5%    ~     (p=0.841 n=5+5)
Tracing/MultinodeCockroach/tracing=t/Insert-32     508µs ± 3%     499µs ± 2%    ~     (p=0.310 n=5+5)

name                                            old alloc/op   new alloc/op   delta
Tracing/Cockroach/tracing=f/Scan1-32              22.0kB ± 2%    21.9kB ± 2%    ~     (p=0.841 n=5+5)
Tracing/Cockroach/tracing=f/Insert-32             35.0kB ± 1%    36.0kB ± 6%    ~     (p=0.730 n=4+5)
Tracing/Cockroach/tracing=t/Scan1-32              29.6kB ± 1%    29.4kB ± 1%    ~     (p=0.222 n=5+5)
Tracing/Cockroach/tracing=t/Insert-32             38.8kB ± 6%    38.5kB ± 4%    ~     (p=0.690 n=5+5)
Tracing/MultinodeCockroach/tracing=f/Scan1-32     65.2kB ± 0%    63.6kB ± 1%  -2.38%  (p=0.016 n=4+5)
Tracing/MultinodeCockroach/tracing=f/Insert-32    77.2kB ± 4%    77.1kB ±12%    ~     (p=0.421 n=5+5)
Tracing/MultinodeCockroach/tracing=t/Scan1-32     82.3kB ± 1%    79.8kB ± 1%  -3.05%  (p=0.016 n=4+5)
Tracing/MultinodeCockroach/tracing=t/Insert-32    80.5kB ± 3%    82.0kB ±24%    ~     (p=0.310 n=5+5)

name                                            old allocs/op  new allocs/op  delta
Tracing/Cockroach/tracing=f/Scan1-32                 246 ± 1%       243 ± 2%    ~     (p=0.397 n=4+5)
Tracing/Cockroach/tracing=f/Insert-32                285 ± 2%       285 ± 1%    ~     (p=0.841 n=5+5)
Tracing/Cockroach/tracing=t/Scan1-32                 269 ± 1%       269 ± 1%    ~     (p=1.000 n=5+5)
Tracing/Cockroach/tracing=t/Insert-32                290 ± 1%       290 ± 1%    ~     (p=0.857 n=5+5)
Tracing/MultinodeCockroach/tracing=f/Scan1-32        768 ± 3%       760 ± 1%    ~     (p=0.833 n=5+5)
Tracing/MultinodeCockroach/tracing=f/Insert-32       679 ± 1%       679 ± 1%    ~     (p=0.667 n=5+4)
Tracing/MultinodeCockroach/tracing=t/Scan1-32        856 ± 2%       855 ± 1%    ~     (p=0.762 n=5+5)
Tracing/MultinodeCockroach/tracing=t/Insert-32       698 ± 0%       704 ± 4%    ~     (p=0.905 n=5+5)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-MjfQqe_4N8N01fE1bbL:b-e7xs4v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[docs/generated/settings/settings-for-tenants.txt, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYqx6l56__mkrU1TFc:-Mk2s2tX5dvuxl3BWyQh:b-utffl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/docs/generated/settings/settings-for-tenants.txt#L108)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Not this PR, but are tenants allowed to define their own collectors? Where do we ensure traces from other tenants don't get collected in this tenant's collector?
</blockquote></details>

discussed offline. Settings are isolated per tenant.
___
*[pkg/util/tracing/grpc_interceptor.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZHOKCBO7IshUjlTjO:-Mk2uKnpD5nE-UYDwk38:b-gq6lss) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/grpc_interceptor.go#L81)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Looks like all we're using this for is to set the error status? Good time to improve the naming for setSpanTags to be a bit more specific. 

[nit] I think I'd also prefer not exporting sp.SetStatus, is it something we expect callers of pkg/tracing to ever use? If not, maybe it's worth just inlining it here.
</blockquote></details>

done and done
___
*[pkg/util/tracing/grpc_interceptor.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYs8vn6X-EhuqkAvdX:-Mk2uRHABzclQZOVvzdx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/grpc_interceptor.go#L321)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Double spaces after //
</blockquote></details>

Done.
___
*[pkg/util/tracing/span.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZIbxvB7fHxveOjsOe:-Mk2uSrLEr7vSkRYDHTi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span.go#L222)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See comment above, I don't think it's that useful to export this outside this package, and maybe worth inlining at caller.
</blockquote></details>

Done.
___
*[pkg/util/tracing/span_inner.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYsu-dAXfiZB65-bKo:-Mk2uwniBVUp7Zuwb6K2:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_inner.go#L150)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/shadowCtx/otelCtx?
</blockquote></details>

done
___
*[pkg/util/tracing/span_options.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYtd2G-DvtUwtZ--A6:-Mk2vpCX099d82Ki7niN:b-u5ryhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_options.go#L31)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Do these symbols need to be exported?
</blockquote></details>

unexported
___
*[pkg/util/tracing/span_options.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYu0BU3884rqgRVnUJ:-Mk2wQzl0pVAa2uFYysB:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/span_options.go#L86)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

"open of a Span"?
</blockquote></details>

fixed
___
*[pkg/util/tracing/tracer.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYuUiJDTyVrgwduQny:-Mk2wvoG2zGv68zz5k-x:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L86)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md

I also thought this wasn't quite the right name for this env var. It should be something like COCKROACH_OTLP_COLLECTOR
</blockquote></details>

done
___
*[pkg/util/tracing/tracer.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYuYQ-9lRGWeQFPXqb:-Mk2xDCV9Nn3Hn-tSztB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L89)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/suppoort/support (+ nit: rephrase to deduplicate support?)
</blockquote></details>

Done.
___
*[pkg/util/tracing/tracer.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYw-Ti-AlncHm2RqfN:-Mk8Je2Y0kAqlEfNTXLl:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L253)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit, throughout] If there's only one kind of shadow tracer now, the OpenTelemetry kind, should we remove all code references to "shadow" tracers? (in favor of otel{Tracers,Span,Ctx,...})
</blockquote></details>

done
___
*[pkg/util/tracing/tracer.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYxnpXD_fCg5m7oAas:-Mk8JrYw92O3QQ2nIfSS:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L348)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This comment can go away.
</blockquote></details>

done
___
*[pkg/util/tracing/tracer.go, line 574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaZFDjL0uvWpV0fzID9:-Mk8KaaV0suM9NcyJtLQ:b-pc9nth) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L574)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Here and below, if there is an error, I think `traceID` would be the zero value. Also below, our error msg captures the traceID, not the spanID. I think printing out the err + value be better.
</blockquote></details>

done (except I'm not printing the value coming out of `ParseUint`; I think it's too trivial).
___
*[pkg/util/tracing/tracer.go, line 631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MaYyOqt-V8ys3hpk8E-:-Mk8KI6FABTshnEgfExx:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a58726a27f4d7d9fa404f1a9fc88224ae7a444cd/pkg/util/tracing/tracer.go#L631)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Nice.
</blockquote></details>

ack


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (approved):
Had a couple of questions and nits but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

> Reviewed 12 of 24 files at r3, 4 of 14 files at r4, 16 of 23 files at r6, 15 of 15 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-MkJ6gO54yr17GAuQFB8:b-uq25kh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @irfansharif)

___
*[pkg/util/tracing/doc.go, line 83 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIyhnd-oVCPzz7zvtU:-MkIyhnd-oVCPzz7zvtV:bxxmyn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/doc.go#L83)):*
> ```Go
> // anyway.
> //
> // The tracing package internally makes use of OpenTelemetry[2] This gives us
> ```

Nit: Add a period after OpenTelemetry[2]
___
*[pkg/util/tracing/grpc_interceptor.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIyrp35MPSb1aQSYjL:-MkIyrp35MPSb1aQSYjM:bw2fe9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/grpc_interceptor.go#L42)):*
> ```Go
> 	// Inject/Extract sense per the OpenTracing spec).
> 	//
> 	// TODO(andrei): The comment above references the OpenTracing spec, but we
> ```

Is this TODO still valid?
___
*[pkg/util/tracing/span.go, line 257 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIzNtrEgEcttxpPQxd:-MkIzNtrEgEcttxpPQxe:bfd0e6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span.go#L257)):*
> ```Go
> 
> func (sm SpanMeta) String() string {
> 	hasShadow := sm.otelCtx.IsValid()
> ```

Nit: Some comments here would be nice around the meaning of `hasShadow` and what the extra functionality entails.
___
*[pkg/util/tracing/span.go, line 260 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIzbVq3O7FStud_G9R:-MkIzbVq3O7FStud_G9S:bqmd8vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span.go#L260)):*
> ```Go
> 	var s strings.Builder
> 	s.WriteString(fmt.Sprintf("[spanID: %d, traceID: %d", sm.spanID, sm.traceID))
> 	if hasShadow {
> ```

Nit: Seems like this `hasShadow` variable is unnecessary? Can't we just do `if sm.otelCtx.IsValid()` directly here?
___
*[pkg/util/tracing/span_options.go, line 21 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ-b9A39hlEs-VNdet:-MkJ-b9A39hlEs-VNdeu:badhobz) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span_options.go#L21)):*
> ```Go
> // spanReferenceType describes the relationship between a parent and a child
> // span. This is an OpenTracing concept that we've maintained back from when
> // CRDB was exporting traces to OpenTrancing collectors. We no longer use
> ```

Nit: typo in OpenTracing.
___
*[pkg/util/tracing/span_options.go, line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ-pzk9TbOFjqExtI6:-MkJ-pzk9TbOFjqExtI7:b-414s2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span_options.go#L122)):*
> ```Go
> // from a parent Span.
> //
> // WithParentAndAutoCollection can be called with a nil sp, in which case it'll
> ```

What is sp? Span?
___
*[pkg/util/tracing/tracer.go, line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ3V5SB962hjoFCIcT:-MkJ3V5SB962hjoFCIcU:b-kcqxrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L296)):*
> ```Go
> 
> 	enableNetTrace.SetOnChange(sv, reconfigure)
> 	openTelemetryCollector.SetOnChange(sv, reconfigure)
> ```

Do we also need to do this for `ZipkinCollector`?
___
*[pkg/util/tracing/tracer.go, line 661 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ4vR-DjcH5swNcTJe:-MkJ4vR-DjcH5swNcTJf:b-8gqo7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L661)):*
> ```Go
> }
> 
> // var noopSpanMeta = (*SpanMeta)(nil)
> ```

Nit: Not related to this PR but seems like this comment can be deleted prior to merge.
___
*[pkg/util/tracing/tracer.go, line 949 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ63FA1RiSMMr5YM0m:-MkJ63FA1RiSMMr5YM0n:broo8aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L949)):*
> ```Go
> 		if localParent != nil {
> 			parentSpanContext = localParent.SpanContext()
> 		} else if remoteParent.IsValid() {
> ```

I'm a bit confused by this. Seems like we are just doing the same thing here as at the start of the method but we aren't using the parentSpanContext anywhere. Am I missing something?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @rimadeodhar)

___
*[pkg/util/tracing/doc.go, line 83 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIyhnd-oVCPzz7zvtU:-MkK_Qw-CQfe8oVVAcD1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/doc.go#L83)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Add a period after OpenTelemetry[2]
</blockquote></details>

done
___
*[pkg/util/tracing/grpc_interceptor.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIyrp35MPSb1aQSYjL:-MkKP4ZG4CIeR5wqs8gp:bqhopnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/grpc_interceptor.go#L42)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Is this TODO still valid?
</blockquote></details>

I've addressed the TODO in that I've deleted the reference to OpenTracing above. I don't think there's anything more to say in this comment.
___
*[pkg/util/tracing/span.go, line 257 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIzNtrEgEcttxpPQxd:-MkKS3RSA9BStdiJBhCK:b-lia94m) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span.go#L257)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Some comments here would be nice around the meaning of `hasShadow` and what the extra functionality entails.
</blockquote></details>

I've added a few more words to the `otelCtx` field in `SpanMeta`.
___
*[pkg/util/tracing/span.go, line 260 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkIzbVq3O7FStud_G9R:-MkKRf6v6ChvVzYE8G1n:b-gwuzr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span.go#L260)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Seems like this `hasShadow` variable is unnecessary? Can't we just do `if sm.otelCtx.IsValid()` directly here?
</blockquote></details>

I've left if for the benefit of the reader, but I've renamed it to `hasOtelSpan` to be more suggestive, and moved it down.
___
*[pkg/util/tracing/span_options.go, line 21 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ-b9A39hlEs-VNdet:-MkKSJLQD3_x5E8uvw1i:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span_options.go#L21)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: typo in OpenTracing.
</blockquote></details>

done
___
*[pkg/util/tracing/span_options.go, line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ-pzk9TbOFjqExtI6:-MkKSQcZ2b6zNQD-cG7R:b-py4aus) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/span_options.go#L122)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

What is sp? Span?
</blockquote></details>

`sp` is the name of the argument. I've put it in quotes to hopefully make that more clear.
___
*[pkg/util/tracing/tracer.go, line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ3V5SB962hjoFCIcT:-MkKSlOo5nUP6a8oaTFn:bogsb27) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L296)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Do we also need to do this for `ZipkinCollector`?
</blockquote></details>

yes, thanks
___
*[pkg/util/tracing/tracer.go, line 661 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ4vR-DjcH5swNcTJe:-MkKSpzv1EhAZqm4wbg7:b32wk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L661)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Not related to this PR but seems like this comment can be deleted prior to merge.
</blockquote></details>

do
___
*[pkg/util/tracing/tracer.go, line 949 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ63FA1RiSMMr5YM0m:-MkKVB7w0sEA2khaeBLR:b-f4moo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L949)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

I'm a bit confused by this. Seems like we are just doing the same thing here as at the start of the method but we aren't using the parentSpanContext anywhere. Am I missing something?
</blockquote></details>

Great catch!
I kept mucking around with this code, and had left it in a non-sensical state. I had been mucking with it because I was hoping to avoid having to allocate a `ctx` to put the parent info in, but unfortunately that's what the otel API forces me to do. I had started conversations in https://github.com/open-telemetry/opentelemetry-go/issues/2251 and https://github.com/open-telemetry/opentelemetry-specification/issues/1943 with the otel people for a better way to pass the parent info; we'll see if we get something.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @rimadeodhar)

___
*[pkg/util/tracing/tracer.go, line 949 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-MkJ63FA1RiSMMr5YM0m:-MkKjGoD04agM6APoUOC:bja2fkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/151f2bc532a46a8b99d28bc472747423e22361b1/pkg/util/tracing/tracer.go#L949)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Great catch!
I kept mucking around with this code, and had left it in a non-sensical state. I had been mucking with it because I was hoping to avoid having to allocate a `ctx` to put the parent info in, but unfortunately that's what the otel API forces me to do. I had started conversations in https://github.com/open-telemetry/opentelemetry-go/issues/2251 and https://github.com/open-telemetry/opentelemetry-specification/issues/1943 with the otel people for a better way to pass the parent info; we'll see if we get something.
</blockquote></details>

I've update `TestOtelTracer` to catch this btw.


<!-- Sent from Reviewable.io -->

andreimatei committed:
vendor: bump modvendor

Pick up the fix for https://github.com/goware/modvendor/issues/25,
allowing us to import packages with uppercase chars in the version name,
such as go.opentelemetry.io/otel v1.0.0-RC3.

Release note: None

andreimatei commented:
I've had to prepend a commit bumping the `modvendor` version to make our vendoring support the funky `-RC3` version of the otel modules. @otan , would you mind stamping that commit pls (and thanks for pointing me to the modvendor issue!)?

otan reviewed (approved):
first commit LGTM

andreimatei reviewed (commented):
TFTRs

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65599#-:-MkZ7Taq6EiYJc8y5bfg:b-6rbzir)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @rimadeodhar)



<!-- Sent from Reviewable.io -->
